[expo-notifications] Add useLastNotificationResponse hook

sjchmiela opened the pull request:
# Why

Idea inspired by @cruzach sounded! https://github.com/expo/expo/pull/10811#pullrequestreview-518819031

# How

- Since there will be two hooks now that depend on the notification response listener being registered in the global scope I moved it to a `SharedNotificationResponseListener.fx.ts` file
- Changed `useInitialNotificationResponse` implementation to rely on it
- Added the ability to listen for incoming responses to the shared listener
- Implemented `useLastNotificationResponse` which is like a notification response swiss knife:
  - if registered early enough judging by its return values states you can deduce whether there was this special "initial notification response" or not
  - always returns the latest notification response received (or something falsy, depending on whether you care about the "initiality" or "initialness" of the response you can ignore the difference between `undefined` or `null` or not)

Frankly speaking I'm not 100% sure how big of a deal is leaving a dangling listener active, but I can't think of any reason why it would be a problem.

# Test Plan

I have verified this hook returns latest notification response if the hook is mounted as (early in the app as possible and after some time from starting the app) ✖️ (the app is started in response to notification tap, notification tap while the app is running)

sjchmiela committed:
[expo-notifications] Add SharedNotificationResponseListener

The listener will be reused by two hooks — useLastNotificationResponse and useInitialNotificationResponse

sjchmiela committed:
[expo-notifications] Add useLastNotificationResponse hook

sjchmiela committed:
[expo-notifications] Add the ability to register for new response to SharedNotificationResponseListener

sjchmiela committed:
[expo-notifications] Add CHANGELOG entry

sjchmiela committed:
[docs][expo-notifications] Add documentation

sjchmiela committed:
[expo-notifications] Remove useInitialNotificationResponse hook

sjchmiela committed:
[expo-notifications] Fetch latest response from native module, resolving from undefined state more quickly

BTW also making EXNotificationsEmitter always listen to events from EXNotificationCenterDelegate

cruzach reviewed (commented):
I tested this (using the code from [this snack](https://snack.expo.io/@charliecruzan/juicy-milkshake)), and I'm consistently getting the initial (i.e. app-launching) notification response!! 🎉  only thing I want to bring up is that I'm seeing some different behavior on iOS vs Android when using **both** `useLastNotificationResponse` and `addNotificationResponseReceivedListener` at the same time. I'm not sure about a use case that would use both, but just because the behavior was different on each platform I thought it might be worth bringing up here:

**iOS**
- only `useLastNotificationResponse` triggered if app is dead & opened with notification
- only `addNotificationResponseReceivedListener` called if app is foregrounded or backgrounded & you tap on a notification

**Android**
- only `useLastNotificationResponse` triggered if app is dead & opened with notification
- both hook AND listener called if app is foregrounded or backgrounded

Sidenote- looks like RN is also sending a warning when you open the app via a notification (this happened to me in development on iOS):
> Sending notification when app is killed, on open: Sending `onDidReceiveNotificationResponse` with no listeners registered. 

Should/could we silence that?

cruzach commented on packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/emitting/NotificationsEmitter.java:
```diff
@@ -47,6 +49,10 @@ public void onDestroy() {
     mNotificationManager.removeListener(this);
   }
 
+  public void getLastNotificationResponseAsync(Promise promise) {
```
should this be an `@ExpoMethod`?

sjchmiela replied:
Of coourse! Thank you for catching that, I was so worked up on iOS not working I forgot to test Android!

cruzach reviewed (commented):
sorry- forgot to add this to the last review

sjchmiela committed:
[expo-notifications] Remove `removeAllNotificationListeners` method which could remove `useLastNotificationResponse` listeners

sjchmiela committed:
Apply suggestions from code review

Co-authored-by: Charlie Cruzan <35579283+cruzach@users.noreply.github.com>

sjchmiela committed:
[expo-notifications] Silence RN warning regression introduced in dfda625f3e7c51fa6c766b60c26dbf09c2987523

sjchmiela commented:
I have now:
- tested it works on Android as expected
- removed the RN warning of "Sending onDidReceiveNotificationResponse with no listeners registered."
- removed `removeAllNotificationListeners()` method — I suspect the dichotomy between Android and iOS with `useLastNotificationResponse` behavior was resulting from `useLastNotificationResponse`'s listener being erroneously removed by the `removeAll…`.

cruzach reviewed (approved):
LGTM 🙌 

Last question- from the looks of it, I don't think any scoped Android code needs to be updated, but `EXScopedNotificationsEmitter.m` does need to be updated so that this fix lands in the managed workflow for SDK 40? I suppose it might make more sense for that to land in a separate PR though (up to you)

sjchmiela committed:
[ios] Fix EXScopedNotificationsEmitter not registering last notification response

sjchmiela commented:
Thank you for catching that! 👁️ 🕵️  https://github.com/expo/expo/pull/10883/commits/3953aa4cd07067221c79062fd9c8ce8fae0b50d5 should take care of it by changing the way we override serialization mechanism and thanks to that allowing us to call `super` in callback methods.

cruzach reviewed (approved):
perfect! thanks for adding that in so quickly 🏎️

SimenB commented:
Sorry to ask here, but is this hook the recommended way of handling notification presses? The guide does not use it, but it's a class component: https://docs.expo.io/push-notifications/receiving-notifications/. It also seems to use the same API as this hook does under the hood.

Currently trying to migrate to `expo-notifications` after pulling out in both SDK 38 and 39, and having issues with presses not triggering our handlers correctly.

cruzach commented:
@SimenB Yes, but this hook landed as part of SDK 40 [(which is currently in beta)](https://blog.expo.io/expo-sdk-40-beta-is-now-available-2e98264f66b2)

You can see documentation for it [here](https://docs.expo.io/versions/v40.0.0/sdk/notifications/#uselastnotificationresponse-undefined--notificationresponse--null), but since it's still in beta i think we're going to keep the example you linked to the same for now 👍

alaminopu commented on packages/expo-notifications/README.md:
```diff
@@ -432,16 +431,16 @@ import * as Notifications from 'expo-notifications';
-      initialNotificationResponse.notification.data.url &&
-      initialNotificationResponse.actionIdentifier === Notifications.DEFAULT_ACTION_IDENTIFIER
+      lastNotificationResponse &&
+      lastNotificationResponse.notification.data.url &&
```
@sjchmiela  I was wondering if this is the correct way to access `notification.data`, since `lastNotificationResponse` contain the response from `addNotificationResponseReceivedListener`. Shouldn't it be `notification.request.content.data` ?

SimenB commented:
> @SimenB Yes, but this hook landed as part of SDK 40 [(which is currently in beta)](https://blog.expo.io/expo-sdk-40-beta-is-now-available-2e98264f66b2)

Yep, I'm testing with the beta 👍 Thanks for the response, 🤞 this works out for us
